src/Database/PG/Query/Connection.hs

```diff
@@ -365,20 +369,48 @@ execPrepared
   :: PQ.Connection                -- ^ connection
   -> [Maybe (DB.ByteString, PQ.Format)]           -- ^ parameters
   -> RemoteKey                    -- ^ stmtName
+  -> ((PQ.Connection -> IO (Maybe PQ.Result)) -> PQ.Connection -> IO (Maybe PQ.Result))
   -> PGExec ResultOk -- ^ result
-execPrepared conn args n = do
-  mRes <- lift $ PQ.execPrepared conn n args PQ.Binary
+execPrepared conn args n run = do
+  mRes <- lift $ run (\c -> PQ.execPrepared c n args PQ.Binary) conn
   checkResult conn mRes
 
+data PGCancelErr = CEAllocate | CEError DT.Text
+  deriving (Show, Eq)
+
+instance Exception PGCancelErr
+
+cancelPG :: PQ.Connection -> IO (Either PGCancelErr ())
+cancelPG conn = do
+  PQ.getCancel conn >>= \case
+    Nothing -> return $ Left CEAllocate
+    Just c  -> PQ.cancel c >>= \case
+      Left err -> return $ Left $ CEError $ lenientDecodeUtf8 err
+      Right () -> return $ Right ()
+
+cancelOnTimeout :: (PQ.Connection -> IO a) -> PQ.Connection -> IO a
+cancelOnTimeout f conn = do
+  a <- async $ f conn
+  x <- (Left <$> wait a) `catch`
+    (\(e::Timeout) -> Right <$> do
```

Comment:
Seems like we're splitting the timeout logic between hge and pg-client here? IIUC we're setting up the exception to be thrown in hge, and catching it here, regardless of whether it was set up to be thrown.

Reply:
These are good points. Again this is very much "try to get something working at all without changing things I don't need to change". I.e., I took the general approach of using `timeout` from graphql-engine as given, and tried to prove that it could be made to work.

Moving the timeout handling to inside pg-client-hs is definitely something to consider -- from an API / encapsulation point of view, it would be great if we didn't have to talk about exceptions in the API docs and could instead allow passing a timeout or deadline. That would just require doing things a bit differently on the graphql-engine level. (I have some vague doubts anyway whether it's a good idea to just throw `Timeout`s at our operations, but those may well be unfounded.)

Reply:
The split of timeout no longer happens in the current version -- we opaquely handle any asynchronous exception as we should :TM:. For the record, I agree that specifying a timeout in the connection settings would probably be a nicer API; the approach I'm taking here is one that fits with using `timeout` higher up in the stack from within graphql-engine.